File: Easy/src/Java/J121/BestTimetoBuyandSellStock.java
package Java.J121;

public class BestTimetoBuyandSellStock {
    public static void main(String[] args) {
        int[] prices1 = {7, 1, 5, 3, 6, 4};
        int[] prices2 = {7, 6, 4, 3, 1};

        BestTimetoBuyandSellStock solution = new BestTimetoBuyandSellStock();
        System.out.println("Example 1: " + solution.maxProfit(prices1)); // Output: 5
        System.out.println("Example 2: " + solution.maxProfit(prices2)); // Output: 0

    }

    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }

        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;

        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else if (prices[i] - minPrice > maxProfit) {
                maxProfit = prices[i] - minPrice;
            }
        }

        return maxProfit;
    }

}

File: Easy/src/Go/Best Time to Buy and Sell Stock.go
package main

import (
	"fmt"
	"math"
)

func maxProfit(prices []int) int {
	minPrice := math.MaxInt64
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
		} else if price-minPrice > maxProfit {
			maxProfit = price - minPrice
		}
	}
	return maxProfit
}

func main() {
	prices1 := []int{7, 1, 5, 3, 6, 4}
	prices2 := []int{7, 6, 4, 3, 1}

	fmt.Println("Пример 1: ", maxProfit(prices1))
	fmt.Println("Пример 2: ", maxProfit(prices2))
}
